pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -24,115 +21,122 @@ var log = logf.Log.WithName("templates")
+	templatesByTier, err := loadTemplatesByTiers(assets)
 	if err != nil {
 		return errors.Wrap(err, "unable to create or update NSTemplateTiers")
 	}
-	tiers, err := g.newNSTemplateTiers(namespace)
+	tierObjs, err := newNSTemplateTiers(s, namespace, templatesByTier)
 	if err != nil {
 		return errors.Wrap(err, "unable to create or update NSTemplateTiers")
 	}
+	// sort the tiers by name so we have a predictive order of creation/update, which also avoids test flakiness
+	tiers := make([]string, 0, len(tierObjs))
+	for tier := range tierObjs {
+		tiers = append(tiers, tier)
+	}
+	sort.Strings(tiers)
 	for _, tier := range tiers {
-		log.Info("creating or updating NSTemplateTier", "namespace", tier.Namespace, "name", tier.Name)
-		if err := client.Create(context.TODO(), tier); err != nil {
-			if !apierrors.IsAlreadyExists(err) {
-				return errors.Wrapf(err, "unable to create the NSTemplateTiers '%s' in namespace '%s'", tier.Name, tier.Namespace)
-			}
-			log.Info("NSTemplateTier resource already exists", "namespace", tier.Namespace, "name", tier.Name)
-			// get the existing NSTemplateTier
-			existing := &toolchainv1alpha1.NSTemplateTier{}
-			err = client.Get(context.TODO(), types.NamespacedName{
-				Namespace: tier.Namespace,
-				Name:      tier.Name,
-			}, existing)
-			if err != nil {
-				return errors.Wrapf(err, "unable to get the NSTemplateTiers '%s' in namespace '%s'", tier.Name, tier.Namespace)
-			}
-			// retrieve the current 'resourceVersion' to set it in the resource passed to the `client.Update()`
-			// otherwise we would get an error with the following message:
-			// "nstemplatetiers.toolchain.dev.openshift.com \"basic\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update"
-			tier.ObjectMeta.ResourceVersion = existing.ObjectMeta.ResourceVersion
-			if err := client.Update(context.TODO(), tier); err != nil {
-				return errors.Wrapf(err, "unable to update the NSTemplateTiers '%s' in namespace '%s'", tier.Name, tier.Namespace)
-			}
-			log.Info("NSTemplateTier resource updated", "namespace", tier.Namespace, "name", tier.Name, "ResourceVersion", tier.ResourceVersion)
+		tierObj := tierObjs[tier]
+		log.Info("creating or updating NSTemplateTier", "namespace", tierObj.Namespace, "name", tierObj.Name)
+		cl := commonclient.NewApplyClient(client, s)
+		created, err := cl.CreateOrUpdateObject(tierObj, true, nil)
+		if err != nil {
+			return errors.Wrapf(err, "unable to create or update the '%s' NSTemplateTiers in namespace '%s'", tierObj.Name, tierObj.Namespace)
+		}
+		if created {
+			log.Info("NSTemplateTier resource created", "namespace", tierObj.Namespace, "name", tierObj.Name)
 		} else {
-			log.Info("NSTemplateTier resource created", "namespace", tier.Namespace, "name", tier.Name)
+			log.Info("NSTemplateTier resource updated", "namespace", tierObj.Namespace, "name", tierObj.Name, "ResourceVersion", tierObj.ResourceVersion)
```

Comment:
the first returned value of `cl.CreateOrUpdateObject` tells if the object was either created or updated - not only created.

Reply:
right, I forgot to update this message after our discussion on BJN

Reply:
fixed in 98dca2f